### ISSUE TITLE
I'm suspicious about part of loss function..

### ISSUE THREAD

**Issue by aif2017** [zzh8829/yolov3-tf2#146]
HI , in part 
true_box_flat = tf.boolean_mask(true_box, tf.cast(obj_mask, tf.bool))
you changed (batch, grid,grid,anchor,4) to (sth,4)
actually you have missed batches because each batch has it's own true_box but you mixed all together  and send it to broadcast_iou(pred_box, true_box_flat).
I think this is a reason of slow training...
thx

**Comment by zzh8829** [comment]
after initial inspection i believe you are right, i am going to do more research and figure out a solution for this. the shape of true_box, obj_mask and true_box_flat are

` [N g g anchor 4] x [N g g anchor] = [nbox 4]  `
but we really want it to be                                                                                                                               
 `   [N g g anchor 4] x [N g g anchor] = [N nbox 4]  `

But it's not easy to do because each image in the batch has different number of labeled boxes
I probably need to tear apart broadcast_iou to fix this. 😢

**Comment by zzh8829** [comment]
The implication of this bug is we are inadequately penalizing false positives for obj which means there would be more false positives in objectiveness. I don't think this error would effect other part of the loss function though.

**Comment by aif2017** [comment]
I think you must change it to image wise loss.

**Comment by zzh8829** [comment]
bug fixed, doesn't seem to make a huge difference in training.
